### ISSUE TITLE
[Enhancement]: Reduce code duplication on segcore

### ISSUE THREAD

**Issue by cydrain** [milvus-io/milvus#17945]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

some suggestions to reduce code duplication on segcore:
1. hafl part of API SegmentSealedImpl::vector_search should be replaced by SearchOnGrowing()
2. FloatSearch and BinarySearch can be merged into one API

**Comment by cydrain** [comment]
current search flowchart is like this:
![search_flow_new](https://user-images.githubusercontent.com/3992404/182323903-e1f0d536-f58f-4c2b-be91-50da84c0bdcd.jpg)

**Comment by cydrain** [comment]
> 1. hafl part of API SegmentSealedImpl::vector_search should be replaced by SearchOnGrowing()
    The first parameter of SearchOnGrowing() is "segcore::SegmentGrowingImpl", it cannot be provided by 
    SegmentSealedImpl::vector_search(), so cannot remove this duplication.
> 2. FloatSearch() and BinarySearch() can be merged into one API
    Merge FloatSearch() and BinarySearch() into SearchOnGrowing()

**Comment by cydrain** [comment]
split SearchOnSealed() into 2 APIs:
- SearchOnSealedIndex() - search on sealed segment index
- SearchOnSealed() -search on sealed segment without index

**Comment by cydrain** [comment]
after #18830, search flowchart is like this
![Flowchart](https://user-images.githubusercontent.com/3992404/186639800-5536b375-a8e3-49bc-87f9-30cbb8993a9c.jpg)
